Switch null assignment to non-nullable entity warnings on

peppy opened the pull request.

peppy committed:
Switch null assignment to non-nullable warnings on

peppy committed:
Fix single case of incorrect usage

bdach committed:
Locally suppress warning in `SerializationReader`

`SerializationReader` is not written in a form that would support
turning nullability checking on for the entire class. The biggest
problem there is the inner `DynamicDeserializer` static class, whose
members are initialised via an `initialize()` method, which the compiler
knows nothing about.

For this reason, just opt to suppress the single inspection about
returning a `null` from a method with a return type of `string` (rider
expects `string?`). It would have been also viable to enable nullability
checking for this one method, but that's pretty much the same thing and
adds no safety anyways, so just disable the warning to minimise
surprise.

bdach committed:
Handle null case explicitly in `SpectatorState.Equals()`

Uses the usual pattern of two `ReferenceEquals` checks against `this`
and `null` before proceeding to inspect field values. Doing this causes
the compiler to infer that at the point that field values are checked,
`other` can no longer viably be `null`.

bdach committed:
Use conditional nullability attribute

As it turns out, C# 8 provides an attribute that allows annotating that
an `out` parameter's nullability depends on the method's return value,
which is exactly what is desired here.

bdach committed:
Remove bogus `InternalChildren` null-check

`InternalChildren` can't viably be `null`, and if it were, we have
bigger problems. The removed null-check was triggering false-positive
inspections further down.

bdach committed:
Pass placeholder hitobject instead of `null`

bdach committed:
Remove list null-checks in `LogoTrackingContainer` test

If the null-checks were tripped, the test would crash anyway. It is not
possible to call `.Any()` and get a valid result instead of an exception
on a null reference.

bdach committed:
Mark access to exception if task faulted as safe

There are seemingly no C#-side compile-time guarantees that it is safe,
but if the task's state is `Faulted` (as is checked right before), the
exception cannot be null as per the documentation.

bdach committed:
Ignore possible nulls in `Type.GetType()` calls

They're mostly used in extensibility scenarios, so everything happens in
runtime. There is no better resolution than to crash with a null
reference exception.

bdach committed:
Ignore possible null in `GetResponseString()`

A null there indicates a deserialisation error and therefore due to the
catch block immediately succeeding the changed line everything will
continue to work as intended.

bdach committed:
Ignore possible null from `JsonConvert.DeserializeObject()`

Nothing better can be done if a `null` is indeed returned.

bdach committed:
Ignore possible path-related nulls

They're all in test code anyway, so any issue there will cause a test to
fail.

bdach committed:
Ignore possible null in multiplayer test

A null value will fail the test anyhow.

bdach committed:
Ignore possible nulls in `NotifyCollectionChangedArgs`

Safe to access by the virtue of the preceding case labels on
`args.Action`.  And they're in test code anyways.

bdach committed:
Ignore possible nulls from stream reader in IPC

Any failures will be caught. They're not logged, but they also weren't
before. Error handling can be improved at a future date, this series of
changes is primarily intending to unblock a new inspection.

bdach commented:
FYI I have a batch of changes that resolves inspections here, just want to run it past tests locally.

bdach commented:
Changes are in. Because of the sheer number of the inspections being fixed the full diff is probably dreadful reviewing, but what I did instead, is I split each individual case/group of cases into separate commits, so I recommend reviewing commit-by-commit, as it'll translate to case-by-case basis. Each commit also has an explanation for the choice undertaken in the expanded description.

I attempted to avoid using `.AsNonNull()`/suppressions where possible, but there was some stuff that was too dynamic to be able to avoid using them entirely.

peppy commented on osu.Game/Skinning/Skin.cs:
```diff
@@ -57,7 +58,7 @@ protected Skin(SkinInfo skin, IStorageResourceProvider resources)
                 string jsonContent = Encoding.UTF8.GetString(bytes);
 
-                DrawableComponentInfo[skinnableTarget] = JsonConvert.DeserializeObject<IEnumerable<SkinnableInfo>>(jsonContent).ToArray();
+                DrawableComponentInfo[skinnableTarget] = JsonConvert.DeserializeObject<IEnumerable<SkinnableInfo>>(jsonContent).AsNonNull().ToArray();
```
I don't think we want to crash as a result of null here (which is why i had the checks above).

bdach replied:
Fair point, tunnel-visioned a bit on that one. Figured that if a `null` falls out of json deserialisation it's a garbage-in-garbage-out scenario already, but I suppose the garbage can just as well be ignored.

bdach replied:
I think this should be better now with 69fc072, if I got the intention right.

peppy commented:
I think the rest seem fine.

bdach committed:
Ignore skin component json data if deserialisation fails instead

Crashing was not really the best thing to do there given the preceding
code that already allowed a few continues in case of a missing file.

bdach committed:
Merge branch 'master' into null-assignment-warning-enable
